src/statsmodels/contrasts.jl

```diff
@@ -0,0 +1,143 @@
+## Specify contrasts for coding categorical data in model matrix. Contrast types
+## are constructed with data and other options, creating a subtype of
+## AbstractContrast. AbstractContrast provides the interface for creating modle
+## matrix columns and coefficient names
+##
+## ModelFrame will hold a Dict{Symbol, T<:AbstractContrast} that maps column
+## names to contrasts.
+##
+## ModelMatrix will check this dict when evaluating terms, falling back to a
+## default for any categorical data without a specified contrast.
+
+abstract AbstractContrast
+
+termnames(term::Symbol, col::Any, contrast::AbstractContrast) =
+    ["$term - $name" for name in contrast.termnames]
+
+function cols(v::PooledDataVector, contrast::AbstractContrast)
+    ## make sure the levels of the contrast matrix and the categorical data
+    ## are the same by constructing a re-indexing vector. Indexing into
+    ## reindex with v.refs will give the corresponding row number of the
+    ## contrast matrix
+    reindex = [findfirst(l .== contrast.levels) for l in levels(v)]
+
+    ## TODO: add kwarg for full-rank contrasts (e.g., when intercept isn't
+    ## specified in a model frame).
+
+    return contrast.matrix[reindex[v.refs], :]
+end
+
+## Default contrast is treatment coding:
+cols(v::PooledDataVector) = cols(v, TreatmentContrast(v))
+termnames(term::Symbol, col::PooledDataArray) = termnames(term, col, TreatmentContrast(col))
+
+
+## Constructing a contrast from a non-pooled data vector will first pool it
+## 
+## (NOT SURE THIS IS GOOD: constructing columns also depends on levels, so for
+## _that_ to work would need to somehow hold onto pooled data...)
```

Comment:
Maybe we could avoid calling `pool` to create the contrasts? Just use `unique`, and allow passing an optional order of levels to `contrast!`. That could also be useful to specify `base` for cases where you want to use a different reference just for one model, without changing it for the PDA itself. IIUC in the current code of this PR, there's no way to pass `base` when calling `ModelFrame`.

You could imagine that `*Contrast` types, instead of actually requiring the data to be passed to their constructors, would just hold the levels in the specified order or the index of the base level, or nothing to keep defaults. Then, under the hood another function would combine the data and the contrast info, and create the contrast matrix.

So you could write things like:
```julia
ModelFrame(Formula(Nothing(), :x), d, contrasts = [:x => SumContrast])
# Equivalent
ModelFrame(Formula(Nothing(), :x), d, contrasts = [:x => SumContrast()])

# With non-default values
ModelFrame(Formula(Nothing(), :x), d, contrasts = [:x => SumContrast(base=2)])
ModelFrame(Formula(Nothing(), :x), d, contrasts = [:x => SumContrast(levels=["A", "C", "B"])])
```
```